common/prometheus.hh

```diff
@@ -0,0 +1,185 @@
+      web_queue_size = &(web_queue_family.Add({}));
+      auto& webhook_queue_family = BuildGauge()
+        .Name(d_prefix+"_webhook_queue_size")
+        .Help("How full is the webhook worker thread queue?")
+        .Register(*d_registry);
+      webhook_queue_size = &(webhook_queue_family.Add({}));
+      commands_family = &(BuildCounter()
+                          .Name(d_prefix+"_commands_total")
+                          .Help("How many commands have been received?")
+                          .Register(*d_registry));
+      custom_family = &(BuildCounter()
+                          .Name(d_prefix+"_custom_stats_total")
+                          .Help("Count of custom stats (by type)")
+                          .Register(*d_registry));
+      dns_query_family = &(BuildCounter()
+                           .Name(d_prefix+"_dns_queries_total")
+                           .Help("How many DNS queries were performed?")
+                           .Register(*d_registry));
+      dns_query_latency_family = &(BuildHistogram()
+                                   .Name(d_prefix+"_dns_query_response_seconds")
+                                   .Help("How long do DNS queries take?")
+                                   .Register(*d_registry));
+      webhook_family =  &(BuildCounter()
+                           .Name(d_prefix+"_webhook_events_total")
+                           .Help("How many webhook events occurred?")
+                           .Register(*d_registry));
+      custom_webhook_family =  &(BuildCounter()
+                                 .Name(d_prefix+"_custom_webhook_events_total")
+                                 .Help("How many custom webhook events occurred?")
+                                 .Register(*d_registry));
+    }
+  }
+  virtual ~PrometheusMetrics() = default;
+  
+  void addCommandMetric(const std::string& name);
+  void incCommandMetric(const std::string& name);
+
+  void addCustomMetric(const std::string& name);
+  void incCustomMetric(const std::string& name);
+
+  void observeWQD(float duration)
+  {
+    worker_queue_duration->Observe(duration);
+  }
+
+  void observeWRD(float duration)
+  {
+    worker_response_duration->Observe(duration);
+  }
+  
+  void setActiveConns(int value)
+  {
+    active_connections->Set(value);
+  }
+
+  void addDNSResolverMetric(const std::string& name);
+  void incDNSResolverMetric(const std::string& name);
+  void observeDNSResolverLatency(const std::string& name, float duration);
+
+  void addWebhookMetric(unsigned int id, const std::string& url, bool custom);
+  void incWebhookMetric(unsigned int id, bool success, bool custom);
+
+  void setWebQueueSize(int value)
+  {
+    web_queue_size->Set(value);
+  }
+  
+  void setWebhookQueueSize(int value)
+  {
+    webhook_queue_size->Set(value);
+  }
+
+  std::string serialize()
+  {
+    return d_textSerializer.Serialize(d_registry->Collect());
+  }
+
+private:
+  Histogram* worker_queue_duration;
```

Comment:
At a quick glance, it looks strange to use raw pointers everywhere.

Reply:
That's because these objects are created and managed by the prometheus Registry class. It returns a reference to the object from BuildXXXX().

Reply:
If you look at Registry.Add() you'll see that the object is managed with a unique ptr